File: packages/i18nify-go/modules/geo/alpha3_to_alpha2.go
package geo

import (
	"fmt"
	"strings"
)

// Alpha3ToAlpha2 converts an ISO 3166-1 alpha-3 country code to its alpha-2 equivalent.
//
// Example:
//
//	Alpha3ToAlpha2("IND")  // → "IN", nil
//	Alpha3ToAlpha2("USA")  // → "US", nil
func Alpha3ToAlpha2(alpha3Code string) (string, error) {
	normalised := strings.ToUpper(strings.TrimSpace(alpha3Code))
	if normalised == "" {
		return "", fmt.Errorf("alpha-3 code is required")
	}

	if cachedCountryMetadata == nil {
		return "", fmt.Errorf("alpha3ToAlpha2: country metadata not loaded")
	}

	for cc, info := range cachedCountryMetadata.GetMetadataInformation() {
		if info != nil && strings.ToUpper(info.GetAlpha_3()) == normalised {
			return cc, nil
		}
	}

	return "", fmt.Errorf("alpha-3 code %q not found", alpha3Code)
}

File: packages/i18nify-js/src/modules/geo/alpha3ToAlpha2.ts
import { withErrorBoundary } from '../../common/errorBoundary';
import { CountryCodeType } from '../types';
import { I18NIFY_DATA_SOURCE } from '../shared';
import { CountryMetaType } from './types';

/**
 * Converts an ISO 3166-1 alpha-3 country code to its alpha-2 equivalent.
 *
 * @param alpha3Code - Three-letter ISO 3166-1 alpha-3 code (e.g. "IND", "USA")
 * @returns {Promise<CountryCodeType>} Resolves to the two-letter alpha-2 code (e.g. "IN", "US")
 */
const alpha3ToAlpha2 = (alpha3Code: string): Promise<CountryCodeType> => {
  const normalised = alpha3Code.trim().toUpperCase();

  if (!normalised) {
    return Promise.reject(
      new Error(
        'Alpha-3 code is required. Please provide a valid ISO 3166-1 alpha-3 code.',
      ),
    );
  }

  return fetch(`${I18NIFY_DATA_SOURCE}/country/metadata/data.json`)
    .then((res) => res.json())
    .then(
      (res: {
        metadata_information: Record<CountryCodeType, CountryMetaType>;
      }) => {
        const entry = Object.entries(res.metadata_information).find(
          ([, info]) => info.alpha_3?.toUpperCase() === normalised,
        );
        if (!entry) {
          return Promise.reject(
            new Error(
              `Alpha-3 code "${alpha3Code}" not found. Please provide a valid ISO 3166-1 alpha-3 code.`,
            ),
          );
        }
        return entry[0] as CountryCodeType;
      },
    )
    .catch((err) => {
      throw new Error(
        `An error occurred while converting alpha-3 to alpha-2. The error details are: ${err?.message || err}.`,
      );
    });
};

export default withErrorBoundary<typeof alpha3ToAlpha2>(alpha3ToAlpha2);
